README.md
# Table of Contents

* [INTRODUCTION] (#introduction)
  * [Overview] (#overview)
  * [Highlight] (#highlight)
  * [Application] (#application)
* [INSTALL] (#install)
  * [Prerequisites] (#prerequisites)
  * [Compile and install] (#compile-and-install)
* [USING GSTREAMILL] (#using-gstreamill)
  * [Prerequisites] (#prerequisites)
  * [B/S Management URL] (#b/s-management-url)
  * [Command line] (#command-line)
  * [Management interface] (#management-interface)
  * [Output] (#output)
* [JOB DESCRIPTION] (#job-description)

# INTRODUCTION

## Overview

gstreamill is an open source, based on gstreamer-1.0 with dvr support and hls output. 

## Highlight

   * hls, http progressive streaminig, udp output.
   * Multi-Rate with GOP Alignment.
   * RESTful management interface, allowing easy integration into operator environment.
   * Job is descript in json.
   * Job run in subprocess, and auto restart on error.
   * Base on gstreamer and easy to extend.
   * B/S Management.
   * Time Shift and Catchup.

## Application

    IP >-----+ 
             |   +------------+   +----> UDP
    CVBS >---+   | gstreamill |   |                 +-- live
             +---+  with dvr  +---+---> M3U8(HLS) --+-- time shift (?timeshift=nnnn)
    SDI >----+   +------+-----+   |                 +-- catch up (?start=yyyymmddhhmmss&end=yyyymmddhhmmss)
             |          |         +---> HTTP
    LIVE >---+          |
                        |
         REST interface (json over http post)

## Third-party Applications

   * Parson
   * JQuery
   * BootStrap
   * BootBox
   * i18next
   * Resumable

# INSTALL

gstreamill have been tested in ubuntu 16.04.2.

## get source

    git clone https://github.com/i4tv/gstreamill.git

## install prerequisites

    sudo apt-get install autoconf automake libtool libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev libgstreamer-plugins-bad1.0-dev libaugeas-dev

## Compile and install

    ./autogen.sh
    ./configure (--help)
    make
    make install

## dist package

    ./make dist

# USING GSTREAMILL

## install prerequisites

    sudo apt-get install gstreamer1.0-plugins-ugly gstreamer1.0-plugins-bad gstreamer1.0-plugins-good gstreamer1.0-plugins-base

## B/S Management URL

    http://gstreamill.ip:20118/admin/

## Command line

* help

    gstreamill -h

        Usage:
          gstreamill [OPTION...]
    
          Help Options:
            -h, --help                        Show help options
            --help-all                        Show all help options
            --help-gst                        Show GStreamer Options
    
          Application Options:
            -j, --job                         -j /full/path/to/job.file: Specify a job file, full path is must.
            -l, --log                         -l /full/path/to/log: Specify log path, full path is must.
            -m, --httpmgmt                    -m http managment service address.
            -a, --httpstreaming               -a http streaming service address.
            -s, --stop                        Stop gstreamill.
            -v, --version                     display version information and exit.

* start gstreamill

    service gstreamill start

* stop gstreamill

    service gstreamill stop

* invoke a job

default management port is 20118, use curl to invoke test job via management interface for example:

        curl -H "Content-Type: application/json" --data @examples/test.job http://localhost:20118/admin/start

* invoke a job in foreground, the debug mode:

        gstreamill -j job_descript_file

* access output via http progressive download:
    
        http://host.name.or.ip:20119/test/encoder/0

* access output use via m3u8:
    
        http://host.name.or.ip:20119/test/playlist.m3u8 (live)
        http://host.name.or.ip:20119/test/playlist.m3u8?timeshift=3600 (time shift)
        http://host.name.or.ip:20119/test/playlist.m3u8?position=1486556428 (time shift Wed Feb  8 20:20:28 CST 2017)
        http://host.name.or.ip:20119/test/playlist.m3u8?start=201506060606&end=20150606070600 (callback)

## Management interface

* start a job over http

        curl -H "Content-Type: application/json" --data @examples/test.job http://localhost:20118/admin/start

        or

        http http://localhost:20118/admin/start < tssegment.job

test.job is job description in json, can be found in examples directory.

* stop a job

        curl http://host.name.or.ip:20118/admin/stop/job_name

job_name is value of the "name" field in job description.

* query gstreamill stat:

        curl http://host.name.or.ip:20118/stat/gstreamill
        curl http://host.name.or.ip:20118/stat/gstreamill/job/test

* query gstreamer information:

        curl http://host.name.or.ip:20118/stat/gstreamer[/plugin]

## Output

job name is the value of name of job description.

* http progressive streaming

        http://host.name.or.ip:20119/job name/encoder/encoder_index

* hls

        http://host.name.or.ip:20119/job name/playlist.m3u8
        http://host.name.or.ip:20119/job name/encoder/encoder_index/playlist.m3u8

* udp

        udp://@ip:port

## JOB DESCRIPTION

Job example:
```json
/*
 * test job, videotestsrc and audiotestsrc
 */
{
    "name" : "test",
    "debug" : "gstreamill:3",
    "source" : {
        "elements" : {
            "videotestsrc" : {
                "caps" : "video/x-raw,width=720,height=576,framerate=25/1"
            },
            "audiotestsrc" : {
                "property" : {
                    "wave" : 8
                }
            }
        },
        "bins" : [
            "videotestsrc ! appsink name=video",
            "audiotestsrc ! appsink name=audio"
        ]
    },
    "encoders" : [
        {
            "elements": {
                "appsrc" : {
                    "property" : {
                        "format" : 3,
                        "is-live" : true
                    }
                },
                "x264enc" : {
                    "property" : {
                        "bitrate" : 1000,
                        "byte-stream" : true,
                        "threads" : 4,
                        "bframes" : 3
                    }
                },
                "faac" : {
                    "property" : {
                        "name" : "faac",
                        "outputformat" : 1
                    }
                },
                "appsink" : {
                    "property" : {
                        "sync" : false
                    }
                }
            },
            "bins" : [
                "appsrc name=video ! queue ! timeoverlay ! queue ! x264enc ! queue ! muxer.",
                "appsrc name=audio ! audioconvert ! audioresample ! voaacenc ! muxer.",
                "mpegtsmux name=muxer ! queue ! appsink"
            ]
        }
    ],
    "m3u8streaming" : {
        "version" : 3,
        "window-size" : 10,
        "segment-duration" : 3.00
    },
    "dvr_duration": 86400
}
```

JOB file structure:

```json
{
    'name' : 'cctv2',
    'debug' : '3',
    'is-live' : false,
    'log-path' : '/home/zhangping/tmp/cctv2',
    'source' : {
        ...
    },
    'encoders' : [
        ...
    ],
    'm3u8streaming' : {
        ...
    },
    "dvr_duration": 86400
}
```

name : job name

debug : debug option, reference gst-launch gst-debug, optional.

is-live : true for live source, false otherwise, for example transcode, default is true.

log-path : dont log to default log direcotry for non-live source, log to log-path if it is presented.

source : source of encoders.

encoders : encoders.

m3u8streaming : m3u8 streaming

dvr_duration : dvr duration, seconds, it's optional.

structure of source:
```json
"source" : {
    "elements" : {
        ...
    },
    "bins" : [
        ...
    ]
}
```
structure of elements:
```json
"elements" : {
    "element" : {
        "property" : {
            ...
        },
        "caps" : "..."
    },
    ...
}
```

structure of bins:
```json
"bins" : [
    bin,
    bin,
    ...
]
```

bins is an array of bin, syntax of bin is like gst-launch, for example:
```json
[
    "udpsrc ! queue ! tsdemux name=demuxer",
    "demuxer.video ! queue ! mpeg2dec ! queue ! appsink name = video",
    "demuxer.audio ! mpegaudioparse ! queue ! mad ! queue ! appsink name = audio" 
]
```
in this example, first bin with tsdemux has sometimes pads, second and third bin link with first bin: demuxer.video and demuxer.audio. second bin with appsink named video and third bin with appsink named audio. source bins must have bin with appsink that is corespond endoders' source.

structure of encoders:
```json
"encoders" : [
    encoder,
    ...
]
```

encoders is an array of encoder, encoder is an object, every encoder correspneds an encode output, that means gstreamill support mutilty bitrate output.

structure of encoder:
```json
{
    "elements" : {
        ...
    },
    "bins" : [
        ...
    ],
    "udpstreaming" : "uri"
}
```

elements and bins is just the same as source structure in syntax, the differnce is encoder bins must have bins with appsrc element, appsrc must have name property, the value of name is the same as appsink name value in source bins. udpstreaming uri is udp streaming output uri, it's optional.

m3u8streaming is hls output, it's optional:
```json
"m3u8streaming" : {
    "version" : 3,
    "window-size" : 10,
    "segment-duration" : 3.00,
}
```

There are examples in examples directory of source.

Talk about gstreamill on [gstreamill](https://groups.google.com/forum/#!forum/gstreamill) or report a bug on [issues](https://github.com/zhangping/gstreamill/issues) page.

## License

gstreamill is available under GPL v3. See the LICENSE file for more info. 


ChangeLog
releases are sorted from youngest to oldest.

v-1.0.0 2019-04-14
- generate segment ahead clock time bug fixed

v-1.0.0-rc6 2018-09-13
- #53 if ((size == -1) && (errno == EAGAIN)) segment fault fixed

v-1.0.0-rc5 2018-08-06
- #53 lock job mutex when reset job

v-1.0.0-rc4 2018-07-31
- #53 lock sempore when clean_job_list

v-1.0.0-rc3 2018-07-30
- #108 check fopen return
- #109 check mmap return

v-1.0.0-rc2 2018-03-15
- add signal number to log message when subprocess terminated by signal

v-1.0.0-rc1 2018
- log rotate bug across year
- timeshift exceed live bug fix
- bandwidth for multi audio tracks

v-0.9.2 2017-10-20
- stream_id 0xea cause tssegment failure bug #103 fix

v-0.9.1 2017-06-26
- removing dvr directory bug fix

v-0.9.0 2017-06-03
- synchronization of master and slave encoder
- codec and bandwidth parameter in m3u8 master playlist
- request time in access log
- memory leak bug fix
- removing dvr directory in a seperate thread

v-0.8.7 2017-04-13
- #97 memory leak bug fix

v-0.8.6 2017-04-09
- gstreamill -s deadlock bugfix
- #97 memory leak bug fix

v-0.8.5 2017-02-24
- log file size 40M
- remove useless log message

v-0.8.4 2017-02-21
- add web style access log file
- use live playlist window size as time shift playlist window size
- time shift position is last segment in time shift playlist
- update parson
- calculate segment duration use buffer pts if no fix_frame_rate eq 0
- dvr download
- position parameter, absolute time shift

v-0.8.3
- check first slice of a picture
- check fps_num zero
- gstreamill -s bug fix
- semaphore deadlock problem

v-0.8.2
- check uri
- h.265 segment support
- record segment in another thread, avoid send segment message fail

v-0.8.1
- tssegment can't segment sichuan local channels

v-0.8.0
- version 0.8.0 release

v-0.8-rc2
- segment only process start failure on gstreamer 1.8.1 fix

v-0.8-RC1
- More meaningful log message
- Calculate segment duration use frame accumulate
- Stop job failure fix

v-0.7.12 2016-01-26
- Add DVR directory managment API
- Deny streaming service bug fix, #76
- Check callback parameters: start & end

v-0.7.11 2015-12-14
- endless loop problem fix: 21bb0df
- remove x264_2pass.log.temp and x264_2pass.log.mbtree.temp file: 892966f5c
- decodebin source example: a91a5f2a0
- response 501 for not impletmented method, for example HEAD
- start job failure bug, due to job semaphore

v-0.7.10 2015-11-20
- segment fault bug fix

v-0.7.9 2015-11-16
- zero size segment

v-0.7.8 2015-11-11
- log rotate in debug mode
- main process log to file in debug mode

v-0.7.7 2015-11-09
- unref job problem
- access job->output->stat in child_watch_cb, cause main crash problem
- -j mode, 0 size segment problem

v-0.7.6 2015-11-05
- increase precision of cpu usage in job stat
- add gstreamill cpu usage in gstreamill stat

v-0.7.5 2015-11-02
- segment only bandwidth parameter in master m3u8 playlist
- add -d command line option, run in foreground for docker compliance

v-0.7.4 2015-10-08
- multi bitrate support

v-0.7.3 2015-09-24
- crash bug fix, #53
- subprocess restart make mistake dvr sequence, #71

v-0.7.2 2015-08-18
- callback parameters buf fix

v-0.7.1 2015-08-14
- dvr bug fix

v-0.7.0 2015-08-04
- add tssegment element
- timeshift parameter: timeshift=nnnn
- callback parameter: start=yyyymmddhhmmss&end=yyyymmddhhmmss
- segment store path: /path/to/dvr/channel_name/encoder_id/yyyymmddhh/
- live hls segment name: yyyymmddhh/us_sequenc_duration.ts
- Dockerfile for container image
- update parson

v-0.6.3 2015-03-04
- bugfix #51

v-0.6.2 2015-02-01
- use autoreconf instead gnome-autogen.sh.
- installation document.

v-0.6.1 2015-01-26
- audio only support.
- vitamio on android player problem at meilisishui.com fix.

v-0.6.0 2015-01-14
- setuid gstreamill, don't run under root now.
- complete media upload, support large media.
- bugfix #11, #12, #30, #31, #33, #35, #36, #38.
- running on CentOS7.
- catch SIGINT and SIGQUIT.
- use unix domain socket instead mqueue #45.

v-0.5.0 2014-09-17 zhangping
- B/S Managment
- Transcode support



src/main.c
/*
 * gstreamill main.
 *
 * Copyright (C) Zhang Ping <dqzhangp@163.com>
 *
 */

#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <glib/gstdio.h>
#include <gst/gst.h>
#include <string.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <locale.h>
#include <pwd.h>
#include <grp.h>
#include <signal.h>

#include "gstreamill.h"
#include "httpstreaming.h"
#include "httpmgmt.h"
#include "parson.h"
#include "jobdesc.h"
#include "tssegment.h"
#include "log.h"

#define GSTREAMILL_USER "gstreamill"
#define GSTREAMILL_GROUP "gstreamill"
#define PID_FILE "/run/gstreamill/gstreamill.pid"

GST_DEBUG_CATEGORY(ACCESS);

GST_DEBUG_CATEGORY(GSTREAMILL);
#define GST_CAT_DEFAULT GSTREAMILL

Log *_log;

static void sighandler (gint number)
{
    _log->log_hd = freopen (_log->log_path, "w", _log->log_hd);
}

static void stop_job (gint number)
{
    GDateTime *datetime;
    gchar *date;

    datetime = g_date_time_new_now_local ();
    date = g_date_time_format (datetime, "%b %d %H:%M:%S");
    GST_WARNING ("\n\n*** %s : job stoped ***", date);
    g_free (date);
    g_date_time_unref (datetime);

    exit (0);
}

/*
 * idle thread for signal SIGTERM - stop the gstreamill.
 * other thread for signal SIGTERM would cause dead lock.
 */
static gpointer idle_thread (gpointer data)
{
    sigset_t set;

    sigemptyset(&set);
    sigaddset(&set, SIGTERM);
    sigaddset(&set, SIGUSR1);
    if (pthread_sigmask (SIG_UNBLOCK, &set, NULL) != 0) {
        g_printf ("sigprocmask failure: %s", strerror (errno));
        exit (-1);
    }

    for (;;) {
        sleep (3600);
    }
}

static void print_version_info ()
{
    guint major, minor, micro, nano;
    const gchar *nano_str;

    gst_version (&major, &minor, &micro, &nano);
    if (nano == 1) {
        nano_str = "(git)";

    } else if (nano == 2) {
        nano_str = "(Prerelease)";

    } else {
        nano_str = "";
    }

    g_print ("gstreamill version: %s\n", VERSION);
    g_print ("gstreamill build: %s %s\n", __DATE__, __TIME__);
    g_print ("gstreamer version : %d.%d.%d %s\n", major, minor, micro, nano_str);
}

#if 0
static gint set_user_and_group ()
{
    struct passwd *pwd;
    struct group *grp;

    /* set group id */
    errno = 0;
    grp = getgrnam (GSTREAMILL_GROUP);
    if ((grp == NULL) && (errno == -1)) {
        perror ("getgrnam failed");
        return 4;

    } else if (grp == NULL) {
        g_print ("getgrnam(\"%s\") failed\n", GSTREAMILL_GROUP);
        return 5;
    }
    if (setgid (grp->gr_gid) == -1) {
        perror ("setuid failure");
        return 6;
    }

    /* set user id */
    errno = 0;
    pwd = getpwnam (GSTREAMILL_USER);
    if ((pwd == NULL) && (errno == -1)) {
        perror ("getpwnam failed");
        return 1;

    } else if (pwd == NULL) {
        g_print ("getpwnam(\"%s\") failed\n", GSTREAMILL_USER);
        return 2;
    }
    if (setuid (pwd->pw_uid) == -1) {
        perror ("setuid failure");
        return 3;
    }

    return 0;
}
#endif

static gint prepare_gstreamill_run_dir ()
{
    gchar *run_dir;
    struct group *grp;

    /* mkdir */
    run_dir = g_path_get_dirname (PID_FILE);
    if (g_file_test (run_dir, G_FILE_TEST_EXISTS)) {
        g_free (run_dir);
        return 0;
    }
    if (g_mkdir (run_dir, 0755) == -1) {
        perror ("chown of gstreamill run directory failure");
        g_free (run_dir);
        return 1;
    }

    /* set own and mod of dir */
    grp = getgrnam (GSTREAMILL_GROUP);
    if ((grp == NULL) && (errno == -1)) {
        perror ("getgrnam failed");
        return 2;

    } else if (grp == NULL) {
        g_print ("getgrnam(\"%s\") failed\n", GSTREAMILL_GROUP);
        return 3;
    }
    if (chown (run_dir, -1, grp->gr_gid) == -1) {
        perror ("chown of gstreamill run directory failure");
        g_free (run_dir);
        return 4;
    }
    if (chmod (run_dir, 0775) == -1) {
        perror ("chmod of gstreamill run directory failure");
        g_free (run_dir);
        return 5;
    }
    g_free (run_dir);

    return 0;
}

static gboolean stop = FALSE;
static gboolean debug = FALSE;
static gboolean version = FALSE;
static gchar *job_file = NULL;
static gchar *log_dir = "/var/log/gstreamill";
static gchar *http_mgmt = "0.0.0.0:20118";
static gchar *http_streaming = "0.0.0.0:20119";
static gchar *shm_name = NULL;
static gint job_length = -1;
static gint shm_length = -1;
static GOptionEntry options[] = {
    {"job", 'j', 0, G_OPTION_ARG_FILENAME, &job_file, ("-j /full/path/to/job.file: Specify a job file, full path is must."), NULL},
    {"log", 'l', 0, G_OPTION_ARG_FILENAME, &log_dir, ("-l /full/path/to/log: Specify log path, full path is must."), NULL},
    {"httpmgmt", 'm', 0, G_OPTION_ARG_STRING, &http_mgmt, ("-m http managment address, default is 0.0.0.0:20118."), NULL},
    {"httpstreaming", 'a', 0, G_OPTION_ARG_STRING, &http_streaming, ("-a http streaming address, default is 0.0.0.0:20119."), NULL},
    {"name", 'n', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_STRING, &shm_name, NULL, NULL},
    {"joblength", 'q', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_INT, &job_length, NULL, NULL},
    {"shmlength", 't', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_INT, &shm_length, NULL, NULL},
    {"stop", 's', 0, G_OPTION_ARG_NONE, &stop, ("Stop gstreamill."), NULL},
    {"debug", 'd', 0, G_OPTION_ARG_NONE, &debug, ("Debug mode, run in foreground."), NULL},
    {"version", 'v', 0, G_OPTION_ARG_NONE, &version, ("display version information and exit."), NULL},
    {NULL}
};

static gint create_pid_file ()
{
    gchar *pid;
    GError *err = NULL;

    pid = g_strdup_printf ("%d", getpid ());
    if (!g_file_set_contents (PID_FILE, pid, strlen (pid), &err)) {
        GST_ERROR ("write pid %s failure: %s\n", PID_FILE, err->message);
        g_error_free (err);
        g_free (pid);
        return 1;
    }
    g_free (pid);

    return 0;
}

static void remove_pid_file ()
{
    if (g_unlink (PID_FILE) == -1) {
        GST_ERROR ("unlink pid file error: %s", g_strerror (errno));
    }
}

Gstreamill *gstreamill;

static void stop_gstreamill (gint number)
{
    /* run in SINGLE_JOB_MODE? just exit */
    if (gstreamill->mode == SINGLE_JOB_MODE) {
        g_printf ("Interrupt signal received\n");
        exit (0);
    }

    /* run in background, stop gstreamill and remove pid file. */
    gstreamill_stop (gstreamill);
    if (number == SIGTERM) {
        remove_pid_file ();
    }
}


static int isrunning_gstreamill()
{
    int ret = 0;

    if (g_file_test (PID_FILE, G_FILE_TEST_EXISTS)) {
        gchar* cmd = g_malloc(512);
        g_sprintf(cmd, "cat %s | head -n 1 | xargs -i ps {} 2>&1 >/dev/null", PID_FILE);
        ret = WEXITSTATUS(system(cmd));

        //exist daemon
        if(ret == 0){
            ret = 1;
        }else{
            g_sprintf(cmd, "rm -f %s", PID_FILE);
            ret = system(cmd);
            ret = 0;
        }

        g_free (cmd);
    }else{
        //do nothing
    }

    return ret;
}

int main (int argc, char *argv[])
{
    HTTPMgmt *httpmgmt;
    HTTPStreaming *httpstreaming;
    GMainLoop *loop;
    GOptionContext *ctx;
    GError *err = NULL;
    gint mode;
    struct rlimit rlim;
    GDateTime *datetime;
    gchar exe_path[512], *date;
    Log *log;

    ctx = g_option_context_new (NULL);
    g_option_context_add_main_entries (ctx, options, NULL);
    g_option_context_add_group (ctx, gst_init_get_option_group ());
    if (!g_option_context_parse (ctx, &argc, &argv, &err)) {
        g_print ("Error initializing: %s\n", GST_STR_NULL (err->message));
        exit (1);
    }
    g_option_context_free (ctx);
    GST_DEBUG_CATEGORY_INIT (ACCESS, "access", 0, "gstreamill access");
    GST_DEBUG_CATEGORY_INIT (GSTREAMILL, "gstreamill", 0, "gstreamill log");

    if (version) {
        print_version_info ();
        exit (0);
    }

    /* stop gstreamill. */
    if (stop) {
        gchar *pid_str;
        gint pid;

        g_file_get_contents (PID_FILE, &pid_str, NULL, NULL);
        if (pid_str == NULL) {
            g_print ("File %s not found, check if gstreamill is running.\n", PID_FILE);
            exit (1);
        }
        pid = atoi (pid_str);
        g_free (pid_str);
        g_print ("stoping gstreamill with pid %d ...\n", pid);
        kill (pid, SIGTERM);
        exit (0);
    }

    /* readlink exe path before setuid, on CentOS, readlink exe path after setgid/setuid failure on permission denied */
    memset (exe_path, '\0', sizeof (exe_path));
    if (readlink ("/proc/self/exe", exe_path, sizeof (exe_path)) == -1) {
        g_print ("Read /proc/self/exe error: %s", g_strerror (errno));
        exit (2);
    }

    if (prepare_gstreamill_run_dir () != 0) {
        g_print ("Can't create gstreamill run directory\n");
        exit (3);
    }
    /*
       if (set_user_and_group () != 0) {
       g_print ("set user and group failure\n");
       exit (4);
       }
     */

    mode = DAEMON_MODE;
    if (job_file != NULL) {
        /* gstreamill command with job, running in single job mode */
        mode = SINGLE_JOB_MODE;

    } else if (debug) {
        /* gstreamill command without job, run in background */
        mode = DEBUG_MODE;
    }

    if (gst_debug_get_default_threshold () < GST_LEVEL_WARNING) {
        gst_debug_set_default_threshold (GST_LEVEL_WARNING);
    }

    /* initialize ts segment static plugin */
    if (!gst_plugin_register_static (GST_VERSION_MAJOR,
                GST_VERSION_MINOR,
                "tssegment",
                "ts segment plugin",
                ts_segment_plugin_init,
                "0.1.0",
                "GPL",
                "GStreamer",
                "GStreamer",
                "http://gstreamer.net/")) {
        GST_ERROR ("registe tssegment error");
        exit (17);
    }

    /* subprocess, create_job_process */
    if (shm_name != NULL) {
        gint fd;
        gchar *job_desc, *p;
        Job *job;
        gchar *log_path, *name;
        gint ret;
        sigset_t set;

        sigemptyset(&set);
        sigaddset(&set, SIGTERM);
        sigaddset(&set, SIGUSR1);
        if (sigprocmask (SIG_UNBLOCK, &set, NULL) != 0) {
            GST_ERROR ("sigprocmask failure: %s", strerror (errno));
            exit (-1);
        }

        /* set subprocess maximum of core file */
        rlim.rlim_cur = 0;
        rlim.rlim_max = 0;
        if (setrlimit (RLIMIT_CORE, &rlim) == -1) {
            GST_ERROR ("setrlimit error: %s", g_strerror (errno));
        }

        /* read job description from share memory */
        job_desc = NULL;
        fd = shm_open (shm_name, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
        if (fd == -1) {
            GST_ERROR ("shm_open error");
            exit (5);
        }
        if (ftruncate (fd, shm_length) == -1) {
            GST_ERROR ("ftruncate error");
            exit (5);
        }
        p = mmap (NULL, shm_length, PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);
        if (p == MAP_FAILED) {
            GST_ERROR ("mmap error: %s", g_strerror (errno));
            exit (5);
        }
        job_desc = g_strndup (p, job_length);

        if ((job_desc != NULL) && (!jobdesc_is_valid (job_desc))) {
            exit (6);
        }

        /* initialize log */
        name = (gchar *)jobdesc_get_name (job_desc);
        if (!jobdesc_is_live (job_desc)) {
            gchar *path;

            path = jobdesc_get_log_path (job_desc);
            log_path = g_build_filename (path, "gstreamill.log", NULL);
            g_free (path);

        } else {
            log_path = g_build_filename (log_dir, name, "gstreamill.log", NULL);
        }

        /* initialize log */
        _log = log_new ("log_path", log_path, NULL);
        g_free (log_path);
        ret = log_set_log_handler (_log);
        if (ret != 0) {
            exit (7);
        }

        /* remove gstInfo default handler. */
        gst_debug_remove_log_function (gst_debug_log_default);

        /* launch a job. */
        datetime = g_date_time_new_now_local ();
        date = g_date_time_format (datetime, "%b %d %H:%M:%S");
        fprintf (_log->log_hd, "\n*** %s : job %s starting ***\n", date, name);
        g_date_time_unref (datetime);
        g_free (date);
        job = job_new ("name", name, "job", job_desc, NULL);
        job->is_live = jobdesc_is_live (job_desc);
        job->eos = FALSE;
        loop = g_main_loop_new (NULL, FALSE);

        GST_INFO ("Initializing job ...");
        if (job_initialize (job, TRUE, fd, p) != 0) {
            GST_ERROR ("initialize job failure, exit");
            exit (8);
        }
        GST_INFO ("Initializing job done");

        GST_INFO ("Initializing job's encoders output ...");
        if (job_encoders_output_initialize (job) != 0) {
            GST_ERROR ("initialize job encoders' output failure, exit");
            exit (8);
        }
        GST_INFO ("Initializing job's encoders output done");

        GST_INFO ("Starting job ...");
        if (job_start (job) != 0) {
            GST_WARNING ("start livejob failure, exit");
            exit (9);
        }
        datetime = g_date_time_new_now_local ();
        date = g_date_time_format (datetime, "%b %d %H:%M:%S");
        fprintf (_log->log_hd, "\n*** %s : job %s started ***\n", date, name);
        g_date_time_unref (datetime);
        g_free (date);
        g_free (name);
        g_free (job_desc);

        signal (SIGPIPE, SIG_IGN);
        signal (SIGUSR1, sighandler);
        signal (SIGTERM, stop_job);

        g_main_loop_run (loop);

    } else {
        /* set parent process maximum of core file */
        rlim.rlim_cur = RLIM_INFINITY;
        rlim.rlim_max = RLIM_INFINITY;
        if (setrlimit (RLIMIT_CORE, &rlim) == -1) {
            GST_ERROR ("setrlimit error: %s", g_strerror (errno));
        }
    }

    /* gstreamill is running? */
    if (isrunning_gstreamill()) {
        g_print ("gstreamill already running !!!\n");
        exit (10);
    }

    /* not SINGLE_JOB_MODE and DEAMON_MODE or DEBUG_MODE? */
    if (mode != SINGLE_JOB_MODE) {
        gchar *path, *log_path, *access_path;
        gint ret;
        sigset_t set;

        sigemptyset(&set);
        sigaddset(&set, SIGTERM);
        if (sigprocmask (SIG_BLOCK, &set, NULL) != 0) {
            g_printf ("sigprocmask failure: %s", strerror (errno));
            exit (-1);
        }
#if 0
        /* pid file exist? */
        if (g_file_test (PID_FILE, G_FILE_TEST_EXISTS)) {
            g_print ("file %s found, gstreamill already running !!!\n", PID_FILE);
            exit (10);
        }
#endif
        /* media directory */
        path = g_strdup_printf ("%s/dvr", MEDIA_LOCATION);
        if (!g_file_test (path, G_FILE_TEST_EXISTS)) {
            g_printf ("Create DVR directory: %s", path);
            if (g_mkdir_with_parents (path, 0755) != 0) {
                g_printf ("Create DVR directory failure: %s", path);
            }
        }
        g_free (path);
        path = g_strdup_printf ("%s/transcode/in", MEDIA_LOCATION);
        if (!g_file_test (path, G_FILE_TEST_EXISTS)) {
            g_printf ("Create transcode directory: %s", path);
            if (g_mkdir_with_parents (path, 0755) != 0) {
                g_printf ("Create transcode directory failure: %s", path);
            }
        }
        g_free (path);
        path = g_strdup_printf ("%s/transcode/out", MEDIA_LOCATION);
        if (!g_file_test (path, G_FILE_TEST_EXISTS)) {
            g_printf ("Create transcode directory: %s", path);
            if (g_mkdir_with_parents (path, 0755) != 0) {
                g_printf ("Create transcode directory failure: %s", path);
            }
        }
        g_free (path);

        /* if DAEMON_MODE or DEBUG_MODE then initialize log */
        if ((mode == DAEMON_MODE) || (mode == DEBUG_MODE)) {
            /* initialize log */
            log_path = g_build_filename (log_dir, "gstreamill.log", NULL);
            access_path = g_build_filename (log_dir, "access.log", NULL);
            log = log_new ("log_path", log_path, "access_path", access_path, NULL);
            g_free (log_path);
            g_free (access_path);
            ret = log_set_log_handler (log);
            if (ret != 0) {
                g_print ("Init log error, ret %d.\n", ret);
                exit (11);
            }

            /* remove gstInfo default handler. */
            gst_debug_remove_log_function (gst_debug_log_default);
        }

        /* if DAEMON_MODE then daemonize */
        if (mode == DAEMON_MODE) {
            /* daemonize */
            if (daemon (0, 0) != 0) {
                fprintf (log->log_hd, "Failed to daemonize");
                remove_pid_file ();
                exit (1);
            }

            /* create pid file */
            if (create_pid_file () != 0) {
                exit (1);
            }
        }

        /* customize signal */
        signal (SIGTERM, stop_gstreamill);

        datetime = g_date_time_new_now_local ();
        date = g_date_time_format (datetime, "%b %d %H:%M:%S");
        GST_WARNING ("\n\n*** %s : gstreamill started ***", date);
        g_free (date);
        g_date_time_unref (datetime);
    }

    /* ignore SIGPIPE */
    signal (SIGPIPE, SIG_IGN);

    loop = g_main_loop_new (NULL, FALSE);

    /* start gstreamill */
    if ((mode == DAEMON_MODE) || (mode == DEBUG_MODE)) {
        gstreamill = gstreamill_new ("mode", mode, "log_dir", log_dir, "log", log, "exe_path", exe_path, NULL);

    } else {
        gstreamill = gstreamill_new ("mode", mode, "log_dir", log_dir, "exe_path", exe_path, NULL);
    }
    if (gstreamill_start (gstreamill) != 0) {
        GST_ERROR ("start gstreamill error, exit.");
        remove_pid_file ();
        exit (12);
    }

    g_thread_new ("idle_thread", idle_thread, NULL);

    /* httpstreaming, pull */
    httpstreaming = httpstreaming_new ("gstreamill", gstreamill, "address", http_streaming, NULL);
    if (httpstreaming_start (httpstreaming, 10) != 0) {
        GST_ERROR ("start httpstreaming error, exit.");
        remove_pid_file ();
        exit (13);
    }

    if (mode != SINGLE_JOB_MODE) {
        /* run in background, management via http */
        httpmgmt = httpmgmt_new ("gstreamill", gstreamill, "address", http_mgmt, NULL);
        if (httpmgmt_start (httpmgmt) != 0) {
            GST_ERROR ("start http mangment error, exit.");
            remove_pid_file ();
            exit (14);
        }

    } else {
        /* run in foreground, start job */
        gchar *job, *p, *result;
        JSON_Value *val;
        JSON_Object *obj;

        /* ctrl-c, stop gstreamill */
        signal (SIGINT, stop_gstreamill);

        /* ctrl-\, stop gstreamill */
        signal (SIGQUIT, stop_gstreamill);

        if (!g_file_get_contents (job_file, &job, NULL, NULL)) {
            GST_ERROR ("Read job file %s error.", job_file);
            exit (15);
        }
        p = gstreamill_job_start (gstreamill, job);
        val = json_parse_string (p);
        obj = json_value_get_object (val);
        result = (gchar *)json_object_get_string (obj, "result");
        GST_INFO ("start job result: %s.", result);
        if (g_strcmp0 (result, "success") != 0) {
            exit (16);
        }
        json_value_free (val);
        g_free (p);
    }

    g_main_loop_run (loop);

    return 0;
}

